### ISSUE TITLE
Issue concerning log when using packing=True

### ISSUE THREAD

**Issue by Forbu** [huggingface/trl#1090]
Hello !
Thanks for the nice work with the TRL lib !
I just want to point a "bug" that can happen when using SFTTrainer (with the option `packing=True`).
Basicly the dataset "pack" the instruction during the iteration (https://github.com/huggingface/trl/blob/a78a05d7b76eb1b4dade9c5c1d4704cceaf8a5b1/trl/trainer/utils.py#L565).
The issue is that it modifies the log step loop of the trainer class 
(https://github.com/huggingface/transformers/blob/ec43d6870aa1afb42a6d2b1b0a03743d3f9b3ce6/src/transformers/trainer.py#L1547)
 that assumes that there is a fix number of step (corresponding to the dataset lenght) for the dataloader (https://github.com/huggingface/transformers/blob/ec43d6870aa1afb42a6d2b1b0a03743d3f9b3ce6/src/transformers/trainer.py#L1824C13-L1824C59)
  especially it you resume the training (like I did in my experiment)
(https://github.com/huggingface/transformers/blob/ec43d6870aa1afb42a6d2b1b0a03743d3f9b3ce6/src/transformers/trainer.py#L1914C77-L1914C91)
Basicly I have two issues using packing=True option :
- first I have a jump in logging epoch values because of the packing : 
```
    {
      "epoch": 0.13,
      "learning_rate": 0.0001,
      "loss": 4.2717,
      "step": 480
    },
    {
      "epoch": 1.0,
      "learning_rate": 0.0001,
      "loss": 4.3144,
      "step": 490
    },
```

- second I encounter a bug with (when I resume my training for the second time after 300x2 steps) :
```There seems to be not a single sample in your epoch_iterator, stopping training at step 600! This is expected if you're using an IterableDataset and set num_steps (11070) higher than the number of available samples.```
There is 11070 steps considering my dataloader batch size and gradient accumulation and the size of my dataset.

I think a simple solution would be to add the `infinite=True` option so I wouldn't have the issue.

I am using deepspeed with 3GPUs.

Version : 
trl : 0.7.4
transformer : 4.36.0

**Comment by lvwerra** [comment]
Hi @Forbu, I believe these issues should have been addressed in https://github.com/huggingface/trl/pull/979. Can you verify?

**Comment by Forbu** [comment]
Hi @lvwerra , I found some workaround (save every epoch rather than a fix number of steps). I can try this with the main version that's it ? I will do some tests this afternoon and tell you
